enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+When reply traffic comes back into the cluster, we must now submit the packet to conntrack to find which network this traffic belongs to. The packet is always first sent into zone 64000, where it is determined whether or not this packet belonged to OVN (CT mark of 1) or the host. Once identified by CT mark as OVN traffic, the packet will then be unSNAT’ed in zone 0 via br-ex rules and the CT mark restored of which network it belonged to. Finally we can send the packet to the correct GR via the right patch port, by matching on the restored CT Mark. From there, OVN will handle unSNAT’ing the masquerade IP and forward the packet to the original pod.
+
+Also to support kubevirt live migration the GR LRP will have an extra address with the configured gateway for the layer2 subnet so the gateway IP is independent of the node where the VM is running on, after live migration OVN should send and GARP for VMs to cleanup its ARP tables since the gateway IP has different mac now.
+
+
+
+Following is a picture that illustrate all these bits with a topology
+
+
+![Layer 2 Egress Topology](images/multi-homing-l2-gw.svg)
+
+
+
+#### Services
+
+When ingress service traffic enters br-ex, there are flows installed that steer service traffic towards the OVN GR. With additional networks, these flows will be modified to steer traffic to the correct GR-&lt;network>’s patch port.
+
+When a host process or host networked pod on a Kubernetes node initiates a connection to a service, iptables rules will DNAT the nodeport or loadbalancer IP into the cluster IP, and then send the traffic via br-ex where it is masqueraded and sent into the OVN GR. These flows can all be modified to detect the service IP and then send to the correct GR-&lt;network> patch port. However, in the case where there is hairpin service traffic to the host (Host->Service->Endpoint is also the host), the endpoint reply traffic will need to be distinguishable on a per network basis. In order to achieve this, each OVN GR’s unique masquerade IP will be leveraged.
+
+For service access towards KAPI/DNS or potentially other services on the default network, there are two potential technical solutions. Assume eth0 is the pod interface connected to the default network, and eth1 is connected to the unique primary network:
+
+1. Add routes for KAPI/DNS specifically into the pod to go out eth0, while all other service access will go to eth1. This will then just work normally with the load balancers on the switches for the respective networks.
+
+2. Do not send any service traffic out of eth0, instead all service traffic goes to eth1. ovn_cluster_router_&lt;unique network> is configured to send service CIDR traffic to ovn-k8s-mp0-&lt;unique network>. ipTables rules in the host only permit access to KAPI/DNS and drop all other service traffic coming from ovn-k8s-mp0-&lt;unique network>. The traffic then gets routed to br-ex and default GR where it hits the service.
```

Comment:
It would not have been DNAT'ed yet, because only the only load balancers configured on the user-defined primary network switch would be the ones that belong to services in that namespace. Therefore for KAPI and DNS, no DNAT will occur in OVN and the packet would come to ovn_cluster_router-<user defined network>. There we can add a route for service cidr traffic to go via ovn-k8s-mp0-<user defined network>.

When the packet comes in from that mp0, we can have iptables rules that drop all traffic destined to the service cidr, except for KAPI/DNS. Now if the packet is destined for KAPI/DNS, it will be routed to br-ex, where it will go towards the OVN default GR that has OVN load balancers on it to DNAT it to the right default cluster endpoint.

I should probably describe this in the document. I'll add it.